File: packages/pvm-assemblyscript/assembly/instructions/system.ts
/**
 * System Instructions
 *
 * ECALLI - Host call with immediate value
 */

import { OPCODE_ECALLI, RESULT_CODE_HOST } from '../config'
import { InstructionContext, InstructionResult } from '../types'
import { BaseInstruction } from './base'

/**
 * ECALLI instruction (opcode 0x0A / 10)
 * Host call with immediate value
 * Dispatches to General functions (0-13) and Accumulate functions (14-26)
 *
 * Gray Paper pvm.tex §7.4.1 line 264:
 * ε = host × immed_X
 *
 * Operand format (lines 251-255):
 * - operands[0:l_X]: immed_X (variable-length immediate, sign-extended)
 * Where: l_X = min(4, ℓ)
 *
 * Note: No encoding byte - just raw immediate bytes (0-4 bytes)
 */
export class ECALLIInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_ECALLI
  name: string = 'ECALLI'

  execute(context: InstructionContext): InstructionResult {
    context.registers[0] = u64(context.operands[0])

    return new InstructionResult(RESULT_CODE_HOST)
  }
}

File: packages/pvm-rust/src/instructions/system.rs
//! System instructions (ECALLI). Mirrors assembly/instructions/system.ts.

use crate::config::{OPCODE_ECALLI, RESULT_CODE_HOST};
use crate::instructions::base::InstructionHandler;
use crate::types::{InstructionContext, InstructionResult};

/// ECALLI (opcode 10): host call with immediate. Returns RESULT_CODE_HOST for PVM to dispatch.
pub struct EcalliInstruction;

impl EcalliInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for EcalliInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_ECALLI)
    }

    fn name(&self) -> &'static str {
        "ECALLI"
    }

    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let host_call_id = context
            .operands
            .first()
            .copied()
            .unwrap_or(0) as u32;
        if let Some(ref mut out) = context.host_call_id_out {
            **out = host_call_id;
        }
        InstructionResult::new(RESULT_CODE_HOST as i32, 0)
    }
}

impl Default for EcalliInstruction {
    fn default() -> Self {
        Self::new()
    }
}
